Demonstrate telemetry

tgross opened the pull request:
@misterbisson this PR demonstrates the new telemetry feature of Containerbuddy.

I've done a bit of work to update the blueprint to use r2 of our Couchbase blueprint before I add the telemetry work, but I've otherwise tried to avoid doing a rework for things like using CNS to bootstrap the connection to Consul.

Changes so far include:
- Our Couchbase r2 does not require bootstrapping via docker exec so eliminate that. But we do need to add polling of the REST API to make sure we wait for CB to be ready before writing indexes.
- Touchbase and Nginx can take advantage of `preStart`
- Minor ports cleanup in the compose files
- Added Prometheus to compose file but its not used in `start.sh` yet
- Wrote sensors and added telemetry stanzas to Nginx config
- Write the sensors and add the telemetry stanzas to Touchbase
- Opened a separate PR to add a telemetry stanza to Couchbase

tgross committed:
Update blueprint for current Containerbuddy and Couchbase

Changes include:
- Our Couchbase r2 does not require bootstrapping via docker exec so eliminate
  that. But we do need to add polling of the REST API to make sure we wait for
  CB to be ready before writing indexes.
- Touchbase and Nginx can take advantage of `preStart`
- Minor ports cleanup in the compose files
- Added Prometheus to compose file but its not used in `start.sh` yet

misterbisson commented on start.sh:
```diff
@@ -219,7 +202,14 @@ setupCouchbase() {
-    removeBucket benchmark
+    while true; do
+        echo -n '.'
+        curl -sf -u ${COUCHBASE_USER}:${COUCHBASE_PASS} \
```
This section appears in progress, but my question is if you're planning to move the bucket and index creation to the Touchbase `preStart` script.

tgross replied:
I was trying to make a minimal intervention in this PR so that we're focusing on telemetry. There's a couple things I'd like to clean up, but I didn't want to go too far down the rabbit hole. Up to you though.

misterbisson replied:
Okay, we both want to change that at some point. Let's see where/how the rest goes.

tgross committed:
Added telemetry support for Nginx

Nginx's telemetry configuration measures its current connections and
makes them available for the Prometheus server. Added a section to our
`./start.sh` script to open the Prometheus graph explorer available.

misterbisson commented on nginx/opt/containerbuddy/nginx.json:
```diff
@@ -16,5 +18,50 @@
+  ],
+  "telemetry": {
+    "name": "telemetry",
+    "url": "/telemetry",
```
Should the URL and name be user-configured options? My reason for asking is how discovery would work in an environment where there may be different for each application.

misterbisson replied:
Proposal: remove `name`, `url`, `poll`, and `ttl` as user-configurable options, so that the `telemetry` stanza is just the following:

``` json
  "telemetry": {
    "port": 9090,
    "sensors": [
      {...}
    ]
  }
```

This goes with https://github.com/joyent/containerbuddy/issues/27#issuecomment-198631596 and the discussion that followed.

tgross replied:
Strictly speaking this is a design question about Containerbuddy's telemetry implementation. So I'm moving this conversation to https://github.com/joyent/containerbuddy/pull/119#issuecomment-206856695 so that anyone following Containerbuddy can follow this discussion.

misterbisson commented on nginx/opt/containerbuddy/nginx.json:
```diff
@@ -16,5 +18,50 @@
+    "url": "/telemetry",
+    "port": 9090,
+    "ttl": 30,
+    "poll": 10,
```
What's the difference between this `poll` and the `poll` inside each `sensor`?

misterbisson replied:
The answer is in https://github.com/tgross/containerbuddy/blob/gh27_metrics/telemetry/README.md#configuring-sensors.

Similar question to the above: is it important to make this a user-configurable setting? Previous conversation: https://github.com/joyent/containerbuddy/issues/27#issuecomment-199255902

tgross replied:
Strictly speaking this is a design question about Containerbuddy's telemetry implementation. So I'm moving this conversation to https://github.com/joyent/containerbuddy/pull/119#issuecomment-206856695 so that anyone following Containerbuddy can follow this discussion.

misterbisson commented on nginx/opt/containerbuddy/sensor.sh:
```diff
@@ -0,0 +1,62 @@
+# have been left here as an example.
+
+# The total number of accepted client connections.
+accepts() {
```
The number of `(${accepts} - ${handled})` is an interesting value. Anything other than `0` is worth setting off alarms for.

Also `(${connections_active} - ${connections_waiting}) / ${worker_connections}` is interesting as well. That will give us a gauge where the closer to `1.0` we get, the harder the instance is working.

tgross replied:
Arguably this is the sort of analysis that you should be doing in the collector -- give Prometheus the raw numbers and do any analysis there. But it's not clear to me whether it's possible to calculate instance-specific derived metrics with Prometheus.

Will add though.

tgross replied:
Pushed up a commit with that change.

tgross committed:
Updated metrics configuration for Nginx

tgross committed:
Update Nginx telemetry config for Containerbuddy 1.4.0-r3

tgross committed:
Add some basic memory metrics to Touchbase

tgross commented:
@misterbisson I've added some simple metrics to Touchbase and opened https://github.com/autopilotpattern/couchbase/pull/15 for Couchbase.

misterbisson commented:
🏡 🚶
